aardvark/updater/__init__.py

```diff
@@ -89,75 +80,69 @@ class property ARN filter
-        - Writes ARNs to a file that phantomjs will read as an input.
-        - Phantomjs exchanges the token for session cookies.
-        - Phantomjs then navigates to the IAM page and executes JavaScript
-        to call GenerateServiceLastAccessedDetails for each ARN.
-        - Every 10 seconds, Phantomjs calls GetServiceLastAccessedDetails
-        - Phantom saves output to a file that is used by `persist()`
-
-        :return: Exit code from phantomjs subprocess32
+        :return: boto3 IAM client in target account & role
         """
-
-        path = os.path.dirname(__file__)
-        console_js = os.path.join(path, 'awsconsole.js')
-
-        with tempfile.NamedTemporaryFile() as f:
-            json.dump(arns, f)
-            f.seek(0)
-            try:
-                p = subprocess32.Popen([
-                    self.current_app.config.get('PHANTOMJS'),
-                    console_js,
-                    token,
-                    f.name,
-                    output_file],
-                    stdout=subprocess32.PIPE, stderr=subprocess32.STDOUT)
-                output, errs = p.communicate(timeout=1200)  # 20 mins
-                self.current_app.logger.debug('Phantom Output: \n{}'.format(output))
-                self.current_app.logger.debug('Phantom Errors: \n{}'.format(errs))
-            except subprocess32.TimeoutExpired:
-                self.current_app.logger.error('PhantomJS timed out')
-                return 1  # return code 1 for timeout
-            except CalledProcessError:
-                self.current_app.logger.error('PhantomJS exited: {}'
-                                              ''.format(p.returncode))
-                return p.returncode
-            else:
-                self.current_app.logger.info('PhantomJS exited: 0')
-                return 0
-
-
-def _get_signin_token(creds):
-    """
-    Exchanges credentials dictionary for a signin token.
-
-    1) Creates URL using credentials dictionary.
-    2) Sends a GET request to that URL and parses the response looking for
-    a signin token.
-
-    :return: Signin Token
-    """
-    url = '{base}?Action=getSigninToken&Session={creds}'
-    url = url.format(base=federation_base_url, creds=creds)
-    return requests.get(url).json()['SigninToken']
+        client = boto3_cached_conn(
+            'iam', account_number=self.account_number, assume_role=self.role_name)
+        return client
+
+    def _call_access_advisor(self, iam, arns):
+        jobs = self._generate_service_last_accessed_details(iam, arns)
+        details = self._get_service_last_accessed_details(iam, jobs)
+        return details
+
+    @staticmethod
+    def _generate_service_last_accessed_details(iam, arns):
+        jobs = {}
+        for role_arn in arns:
+            job_id = iam.generate_service_last_accessed_details(Arn=role_arn)['JobId']
+            jobs[job_id] = role_arn
+        return jobs
+
+    def _get_service_last_accessed_details(self, iam, jobs):
```

Comment:
```suggestion
    def _get_service_last_accessed_details(self, iam, jobs):
        access_details = {}
        job_queue = list(jobs.keys())
        last_job_completion_time = time.time()
        while job_queue:
            now = time.time()
            if now - last_job_completion_time > self.max_access_advisor_job_wait:
                # We ran out of time, some jobs are unfinished
                self._log_unfinished_jobs(job_queue, jobs)
                break
            job_id = job_queue.pop()
            role_arn = jobs[job_id]
            details = iam.get_service_last_accessed_details(JobId=job_id)
            if details['JobStatus'] == 'IN_PROGRESS':
                job_queue.append(job_id)
                if not job_queue:  # We're hanging on the last job, let's hang back for a bit
                    time.sleep(1)
                continue
            if details['JobStatus'] != 'COMPLETED':
                self.current_app.logger.error(
                    "Job {job_id} finished with unexpected status {status} for ARN {arn}.".format(
                        job_id=job_id,
                        status=details['JobStatus'],
                        arn=role_arn,
                ))
                continue
            last_job_completion_time = time.time()
            access_details[role_arn] = details['ServicesLastAccessed']
            for detail in access_details[role_arn]:
                last_auth = detail.get('LastAuthenticated')
                if last_auth:
                    last_auth = time.mktime(last_auth.timetuple())
                else:
                    last_auth = 0
                detail['LastAuthenticated'] = last_auth
        return access_details
```

The 5 minute timeout was causing us problems on our larger accounts.  I switched it so that it resets the timeout everytime we get a completed job back.

Should just be three lines that are changed:

```
$ diff __init__.py other.py
107c107
<         last_job_completion_time = time.time()
---
>         start_time = time.time()
110c110
<             if now - last_job_completion_time > self.max_access_advisor_job_wait:
---
>             if now - start_time > self.max_access_advisor_job_wait:
130d129
<             last_job_completion_time = time.time()
```

See any problems with this change?

Reply:
No issues with the change, looks good to me. If we are changing to this approach, I wonder if 5 minutes is too long to wait? It essentially means we give a job 5 minutes _after_ everything else completed? Not a big issue but I wonder whether a shorter timeout would lead to the same result?

I'm surprised by the impact you're seeing though. I would have assumed that the jobs get processed in parallel but it looks like we're facing per-account job processing, so more jobs means longer time until the last job is finished which I wouldn't have expected.

Reply:
For testing I'm only running it against one account, but that account just happens to be ginormous, with thousands and thousands of ARNs to look at.  Took 15-20 minutes to generate the job-ids, and then probably another 7-12 minutes for all the jobs to come back as COMPLETED.

I don't think the 5 minutes will be a mandatory delay after everything else completes because the job_queue will be empty once everything completes so the `while` loop will exit.  5 minutes would be a worst case scenario while waiting for a specific job to go from IN_PROGRESS to COMPLETED.

Could probably be lowered, but I don't see too much harm in keeping it at 5 min since it doesn't commit us to waiting that long.

Reply:
Oh interesting, yeah I didn't consider that case. So if simply running through the loop takes more than 5 minutes then the old approach won't work. I like your approach 👍